File: leap-frontend/src/app/core/services/comment.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { CommentDto, CommentThreadDto, CreateCommentRequest } from '../../shared/models/comment.model';

@Injectable({ providedIn: 'root' })
export class CommentService {

  private readonly baseUrl = '/api/comments';

  constructor(private http: HttpClient) {}

  getThread(reportType: string, lineKey: string, segmentName: string | null): Observable<CommentThreadDto[]> {
    let params = new HttpParams()
      .set('reportType', reportType)
      .set('lineKey', lineKey);
    if (segmentName) {
      params = params.set('segmentName', segmentName);
    }
    return this.http.get<CommentThreadDto[]>(this.baseUrl, { params });
  }

  getHierarchyThreads(reportType: string, segmentName: string, lineKey: string): Observable<Record<string, CommentThreadDto[]>> {
    const params = new HttpParams()
      .set('reportType', reportType)
      .set('segmentName', segmentName)
      .set('lineKey', lineKey);
    return this.http.get<Record<string, CommentThreadDto[]>>(`${this.baseUrl}/hierarchy`, { params });
  }

  create(request: CreateCommentRequest): Observable<CommentDto> {
    return this.http.post<CommentDto>(this.baseUrl, request);
  }

  update(id: number, content: string, driverCode?: string): Observable<CommentDto> {
    return this.http.put<CommentDto>(`${this.baseUrl}/${id}`, { content, driverCode });
  }

  delete(id: number): Observable<void> {
    return this.http.delete<void>(`${this.baseUrl}/${id}`);
  }

  getCounts(reportType: string, segmentName: string | null): Observable<Record<string, number>> {
    let params = new HttpParams().set('reportType', reportType);
    if (segmentName) {
      params = params.set('segmentName', segmentName);
    }
    return this.http.get<Record<string, number>>(`${this.baseUrl}/counts`, { params });
  }
}

File: leap-comment/src/main/java/com/leappoc/comment/service/CommentService.java
package com.leappoc.comment.service;

import com.leappoc.shared.dto.*;

import java.util.List;
import java.util.Map;

public interface CommentService {

    List<CommentThreadDto> getThread(String reportType, String lineKey, String segmentName, String currentUserId);

    Map<String, List<CommentThreadDto>> getHierarchyThreads(String reportType, String segmentName,
                                                             String lineKey, String currentUserId);

    CommentDto createComment(CreateCommentRequest request,
                             String userId, String displayName, String email);

    CommentDto updateComment(Long id, String content, String driverCode, String currentUserId);

    void deleteComment(Long id, String currentUserId, boolean isAdmin);

    Map<String, Long> getCounts(String reportType, String segmentName);
}
